docs/aztec3/protocol/communication-abstractions.mdx

```diff
@@ -189,38 +194,225 @@ struct L2L1CrossMsg {
+Say the public state holds a `mapping(address user => bool blacklisted)` and a value with the block number of the last update `last_updated`. The private functions can then use this public blacklist IF it also performs a public function call that reverts if the block number of the historic state is older than the `last_updated`. This means that updating the blacklist would make pending transactions fail, but allow a public blacklist to be used. Similar would work for the Aave example, where it is just a public value with the allowed caller contracts. Example of how this would be written is seen below. Note that because the `onlyFresh` is done in public, the user might not know when he is generating his proof whether it will be valid or not. 
 
-```
+```solidity
 function transfer(
-    secret address to,
-    secret uint256 amount,
-    secret HistoricState state
-    ) secret returns(bool) {
-    // Some spoonfeeding of proofs that check
-    if (blacklisted[msg.sender] && isFresh(state.blockNumber)) {
-        revert();
-    }
+  secret address to,
+  secret uint256 amount,
+  secret HistoricState state
+  ) secret returns(bool) {
+  if (blacklisted[msg.sender] || blacklisted[to]) revert("Blacklisted");
+  onlyFresh(state.blockNumber);
+  // logic    
 }
 
-function isFresh(pub uint256 blockNumber) public returns(bool){
-    return blockNumber > last_updated;
+function onlyFresh(pub uint256 blockNumber) public {
+  if (blockNumber < last_updated) revert("Stale state");
 }
 ```
 
+:::info
+This is not a perfect solution, as any functions using access control might end up doing a lot of public calls it could put a significant burden on sequencers and greatly increase the cost of the transaction for the user. We are investigating ways to improve.
+:::
+
+
+## Linking L1 and L2 contracts
+As mentioned earlier, there will be a link between L1 and L2 contracts (with the L1 part of the link being the portal contract), this link is created at "birth" when the contract leaf is inserted. However, the specific requirements of the link is not yet fully decided. And we will outline a few options below. 
+
+The reasoning behind having a link, comes from the difficulty of L2 access control (see "A note on L2 access control"). By having a link that only allow 1 contract (specified at deployment) to send messages to the L2 contract makes this issue "go away" from the application developers point of view as the message could only come from the specified contract. The complexity is moved to the protocol layer, that must now ensure that messages to the L2 contract are only sent from the specified L1 contract. 
```

Comment:
You might wanna add that this only solves the access control cases of the right L1 contract sending messages